### ISSUE TITLE
match on a tuple does not properly exhaust/narrow cases

### ISSUE THREAD

**Issue by SebbyLaw** [microsoft/pyright#6437]
**Describe the bug**
A match statement on a tuple of values will not narrow properly even if all valid cases are satisfied.

**Code or Screenshots**

With a bool and a literal:
```python
from typing import Literal

def fn(foo: bool, bar: Literal["a", "b"]) -> int:
    match foo, bar:
        case True, "a":
            return 1
        case True, "b":
            return 2
        case False, "a":
            return 3
        case False, "b":
            return 4
```
Error:
```
Cases within match statement do not exhaustively handle all values
  Unhandled type: "tuple[bool, Literal['a', 'b']]"
  If exhaustive handling is not intended, add "case _: pass"Pylance[reportMatchNotExhaustive](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportMatchNotExhaustive)
```
However, at runtime:
```python
# these work as intended
print(fn(True, "a"))
print(fn(True, "b"))
print(fn(False, "a"))
print(fn(False, "b"))
```
This is not special to the data types `boolean` and `Literal`, any combination of the two behave the same way.

Duplicate cases aren't reported as well:
```python
from typing import Literal

def fn(foo: bool, bar: Literal["a", "b"]) -> int:
    match foo, bar:
        case True, "a":
            return 1
        case True, "b":
            return 2
        case False, "a":
            return 3
        case False, "b":
            return 4
        case True, "a":  # should be: Pattern will never be matched for subject type "Never" (PylancereportUnnecessaryComparison)
            # does not actually get narrowed by the earlier case
            return 5
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
pyright 1.1.335

**Comment by erictraut** [comment]
Yes, this is expected behavior, so I don't consider it a bug.

The type of the subject expression `(foo, bar)` is `tuple[bool, Literal["a", "b"]]`. When narrowing the types of tuples in a `match` statement, pyright narrows the type in the "fall through" case only if narrowing affects a single element of tuple. To do otherwise would result in a combinatoric explosion of tuple types.

For example, the fall-through type after the first case statement would need to be `tuple[bool, Literal["b"]] | tuple[Literal[False], Literal["a", "b"]]`. If the tuple length is `n`, this would result in `n!` combinations. This quickly becomes impractical, so I needed to pick a cutoff point. I decided to support only the case where a single tuple entry is narrowed by a case statement.

For example, this works fine:
```python
def fn(foo: bool, bar: Literal["a", "b"]) -> int:
    match foo, bar:
        case True, x:
            return 1 if x is "a" else 2
        case False, x:
            return 3 if x is "a" else 4
```

I'll note that other Python type checkers like mypy don't attempt to do any narrowing if the subject expression is a tuple, so pyright is already going well beyond what other type checkers do here.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
